axlearn/common/utils.py

```diff
@@ -830,9 +832,13 @@ def cast_floats(in_tree: NestedTensor, to_dtype: Optional[jnp.dtype]) -> NestedT
 
     from_dtype = jnp.float32 if to_dtype == jnp.bfloat16 else jnp.bfloat16
 
-    def cast(x: Tensor) -> Tensor:
+    def cast(x: Union[Tensor, TensorSpec]) -> [Tensor, TensorSpec]:
         if x.dtype == from_dtype:
-            return x.astype(to_dtype)
+            if isinstance(x, Tensor):
+                return x.astype(to_dtype)
+            else:
+                x.dtype = to_dtype
+                return x
```

Comment:
To avoid mutating `x`:

```suggestion
                return dataclasses.replace(x, dtype=to_dtype)
```

Reply:
changed